### ISSUE TITLE
allow voluntary .ts suffix for import paths

### ISSUE THREAD

**Issue by timreichen** [microsoft/TypeScript#37582]
## Search Terms
.ts
suffix
imports
extension

## Suggestion
Typescript doesn't recognize file imports with ```.ts``` suffix.
Allow voluntary ```.ts``` to be added to import paths.
## Use Cases

It seems right to be able to use a correct path to a file without magic resolution.
This would help to align with [deno](https://deno.land) which uses mandatory suffixes o files.

## Examples

let
```import a from "path/to/a.ts"```
behave the same as 
```import a from "path/to/a"```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by timreichen** [comment]
After some more research I wonder if that behavior could even be more guided by a compiler option like ```--noImplicitSuffix```.
That would be in alignment with ```--noImplicitAny``` and give the developer the choice, yet push for a valid, not "magical" resolved path.

What are your thoughts on that?

**Comment by Jack-Works** [comment]
#35148

**Comment by zacnomore** [comment]
> #35148
Read through the above and it seems to boil down to multiple build targets with different file extensions throws a wrench in things when supporting using the file extensions on imports. Where it diverges in my mind, and hopefully @Jack-Works you can illuminate this a bit, is that we'd be importing .ts extensions which would have to be rewritten anyways.

**Comment by resynth1943** [comment]
So... where are we *actually* at with this?

Scattered issues / feature requests don't help, and the confusing (and somewhat untimely) response(s) from official TypeScript members haven't helped me understand *how* this feature will be implemented (it will have to be, eventually).

All I can ascertain is that this is has yet to be resolved.

Can someone explain *in plain English** when and how these multiple issues will be fixed, please?

**Comment by tonygiang** [comment]
+1

If I turn on `resolveJsonModule` and put a `MatchmakingConfig.ts` and a `MatchmakingConfig.json` in the same path, a nasty unexpected behavior awaits:

`import MatchmakingConfig from "./path/to/MatchmakingConfig"; //this results in MatchmakingConfig.json being imported, not MatchmakingConfig.ts`!

Allowing `.ts` suffix will let us specify `.ts` files specifically.

**Comment by m93a** [comment]
I would love to see this issue resolved. Without the support for imports with a `.ts` extension, it's nearly impossible to create a repository that works with Deno, has TS language services and is buildable with `tsc`. It seems like an arbitrary limitation on the TypeScript side 😕️

**Comment by cztomsik** [comment]
@m93a  I am using esbuild for transpilation/bundling and the resulting bundle then works in both deno and node. It's a workaround but at least it works.

**Comment by patrickarlt** [comment]
This would also help for things like https://github.com/import-js/eslint-plugin-import/issues/2111#issuecomment-915663474. Right now `eslint-plugin-import` thinks that TypeScript files should only import `.ts` files or omit the extension.

**Comment by Garcia101** [comment]
> Typescript doesn't recognize file imports with `.ts` suffix.
> Allow voluntary `.ts` to be added to import paths.

JavaScript doesn't differentiate between file extensions. It simply doesn't care. TypeScript should do the same, by simply accepting any file, and using it's MIME type instead.

**Comment by cawoodm** [comment]
This is pretty crazy with Deno:
* `import { Foo } from "./types.ts";` -> Runs fine but VS Code shows a syntax error:
  * `An import path cannot end with a '.ts' extension. Consider importing './types' instead.ts(2691)`
* `import { Foo } from "./types";` -> Does not run but VS Code is happy:
  * `error: Cannot resolve module /my/path/to/definitely/existing/types`

**Comment by andrewbranch** [comment]
I think you’re right that this would be fairly easy from an implementation standpoint, but there are two main reasons why we don’t want to move forward without careful consideration of the full problem space:

1. If we _simply_ add a flag that allows module resolution to work with `.ts` files, it will 100% contribute to the misconception that TypeScript does, can, or should emit JS files with the transformation `import "./foo.ts"` → `import "./foo.js"`. We already have angry mobs demanding that this happen for `import "./foo"` → `import "./foo.js"` so if we allowed `.ts` extensions on there without it being very clear what use cases it is intended to serve, it would just pour fuel on that fire.
2. Such a flag would look like a gift to Deno users, but it would actually be woefully incomplete for them. There are obviously a lot of other parts of module resolution that are either specific to Deno or shared between Deno and the browser that we don’t currently support, and I think those users would be confused and disappointed that we stopped so far short of proper support for these things.

Don’t get me wrong, I want us to solve all of these problems, but we’re still in the phase where we’re trying to develop a thorough understanding of all the different reasons people want this and what the existing solutions are.

**Comment by JasonKleban** [comment]
> I’m not convinced that node12/nodenext are appropriate module resolution modes if you’re writing for the browser or for a bundler.

What about for isomorphic server-side-rendered code?

**Comment by andrewbranch** [comment]
That’s a good question. You would clearly need to write that code using the lowest common denominator of resolution features supported by all of your targets. The combinatorics of this mean that such a mode will probably never be supported by TypeScript, so you’d have two options:

- Use a single TypeScript module resolution mode that supports _at least_ all of those features and use caution and/or third-party tooling to ensure you don’t use the features that won’t work in all targets.
- If there is a good TypeScript module resolution mode for _each_ of your targets, check the same code multiple times, once per target module resolution mode, to ensure it will work in each target.

There are some combinations of targets that seem to have no possible overlap. For example, if you want to write a relative import path to a TS file in Deno, you need to include the `.ts` extension. But if you want to compile that same code to work in the browser, you might be stuck, at least if you use `tsc` alone, as our current thinking is that if we let you write `.ts` file extensions we would force you to compile with `--noEmit`. Of course, there are a lot of other tools that could come into play and help you out here. [Vite](https://vitejs.dev) seems particularly well-positioned to help with situations like these since it’s ESM-first and transpiles TS without checking—this means you could satisfy our type checker by telling it `--noEmit` while Vite handles your emit through ESBuild.

But overall, I think we probably need to start by solving the most common individual scenarios first before we figure out how they can be endlessly combined.

**Comment by cawoodm** [comment]
> @cawoodm do you have the Deno extension for VS Code enabled? https://deno.land/manual@v1.14.3/vscode_deno
> My understanding is that the IDE experience with Deno is pretty well solved...
I did indeed but there were teething issues which have since disappeared. `import { Foo } from "./types.ts";` is now accepted.

**Comment by thescientist13** [comment]
Just wanted to chime as guided from [my comment in the other thread](https://github.com/microsoft/TypeScript/issues/27481#issuecomment-976665951) and alluded to with Vite here and [Snowpack](https://github.com/microsoft/TypeScript/issues/27481#issuecomment-745750281), lots of tools besides Deno want to provide first class support for TS, it certainly stands as a testament to the great work of the team!  But all of this is being made possible via ESM, but the ESM spec requires an extension.  So in an ESM based world, omitting the extension is definitely going against the grain, and I expect it will be so more and more going forward I suspect.

So in the iterium, the current choices seem to be:
- no extension
- or use .js

Which leaves those wanting to provide first class support for TS between a rock and a hard place.  

----

On another, note, perhaps this is something that could be reconciled through [import assertions](https://github.com/tc39/proposal-import-assertions)?  Now that [JSON and CSS are loadable via ESM](https://www.bram.us/2021/10/14/esnext-import-assertions-json-modules-css-modules/), perhaps this could be a spec compliant work around?  🤔 
```ts
import { Foo } from "./foo.js" assert { type: "ts" };
```

Something that gives a bit of hinting for both language and tooling authors alike?

**Comment by jespertheend** [comment]
> but we’re still in the phase where we’re trying to develop a thorough understanding of all the different reasons people want this

There's two reasons why I want this:
- Type checking `.js` files that run in the browser without any build step. All source files are in pure `.js` and all types are provided with JSDoc. With `"checkJs": true` and `"strict": true` types can be used for completions and for continuous integration this way.
- Type checking js/ts files that run in Deno. Ideally the Deno extension would solve this, but using the extension comes with a few other issues. Mainly https://github.com/denoland/deno/issues/13189 and https://github.com/denoland/vscode_deno/issues/511 but a few other ones as well.

**Comment by donmccurdy** [comment]
I'm finding this issue to be the single biggest obstacle to writing a TypeScript library that works in all environments (Web, Node, and Deno). I'm compiling the library to a single file with [microbundle](https://github.com/developit/microbundle). Ideally I could run the same test suite across both Node and Deno...

```shell
ts-node tape tests/*.test.ts

deno test --import-map tests/deno-import-map.json --no-check tests/
```

... but the supported import statements for the two environments are mutually exclusive. 😕

**Comment by andrewbranch** [comment]
@jespertheend can you elaborate on the first of your two bullet points? I don’t understand how that one is related to writing `.ts` in import statements. If all your files are `.js` to begin with, everything should work with `.js` in import statements, no?

**Comment by jespertheend** [comment]
Ah sorry, for a second I thought this issue was about url imports as well as the `.ts` extension.
When I wrote that I was trying to import `{assertEquals} from "https://deno.land/std@0.118.0/testing/asserts.ts";`, which would still suffer from `Cannot find module. Did you mean to set the 'moduleResolution' option to 'node', or to add aliases to the 'paths' option? ts(2792)` even if `.ts` imports are allowed.

**Comment by Lilja** [comment]
> If we simply add a flag that allows module resolution to work with .ts files, it will 100% contribute to the misconception that TypeScript does, can, or should emit JS files with the transformation import "./foo.ts" → import "./foo.js". We already have angry mobs demanding that this happen for import "./foo" → import "./foo.js" so if we allowed .ts extensions on there without it being very clear what use cases it is intended to serve, it would just pour fuel on that fire.

I would like to state that the argument for `./foo` → `./foo.js` is not nearly as strong as for `./foo.ts` → `./foo.js`. The `.ts`/`.d.ts` file extension very explicitly adds which file that is indented to import. In contrast to `./foo` which could mean `./foo.ts`, `./foo/index.ts`, `./foo/index.d.ts`. `./foo` is a lot more ambiguous, while `./foo.ts` is a lot more explicit.

I'm all for removing the amount of ambiguity so I'll gladly help the typescript compiler point to which file I'm actually referring to and if that in turn will make the typescript compiler turn those file extensions into `.js` to help with ESM compatibility, then it's terrific.
Explicit is almost always better than implicit, I wouldn't lose sleep over people who still want `./foo` -> `./foo.js` if `*.ts` was resolved to `.js` while files without file extensions wouldn't.

**Comment by arendjr** [comment]
@andrewbranch 

> If we simply add a flag that allows module resolution to work with .ts files, it will 100% contribute to the misconception that TypeScript does, can, or should emit JS files with the transformation import "./foo.ts" → import "./foo.js". We already have angry mobs demanding that this happen for import "./foo" → import "./foo.js" so if we allowed .ts extensions on there without it being very clear what use cases it is intended to serve, it would just pour fuel on that fire.

FWIW, I was not aware of this issue when I filed https://github.com/microsoft/TypeScript/issues/49083, but it appears I’m now part of the angry mob you refer to :)

I appreciate you seem to be willing to actually work on this, but unfortunately it seems we’re not actually getting close to a solution. Meanwhile the `node16` module resolution is about to be released which is also pouring fuel on that same fire in its own way.

As @Lilja states, support doesn’t need to extend beyond explicit, full paths. But it would really be great if we could *some* support for this.

**Comment by arendjr** [comment]
As for @andrewbranch’s other point:

> Such a flag would look like a gift to Deno users, but it would actually be woefully incomplete for them. There are obviously a lot of other parts of module resolution that are either specific to Deno or shared between Deno and the browser that we don’t currently support, and I think those users would be confused and disappointed that we stopped so far short of proper support for these things.

This seems to be a classic defeatist argument. Just because you cannot solve *all* problems should not be an argument to not solve *any*.

**Comment by inca** [comment]
First, let me start with a (hopefully) constructive suggestion: let's please remove the "support extensionless imports" from the main scope and stop discuss it alongside the "add support for importing `.ts` extension". In my opinion it adds unnecessary controversy to any extension-related problem — and it is more or less clear that probing for different extensions is not feasible for numerous of platforms like Web and Deno (but not limited to those two). So I suggest we stop mixing these two problems and assume that we do not want to write extensionless imports and have TSC magically guess what we meant.

---

Now, on "import `.ts`" thing: since I've read a lot of different interpretations of "how relative imports in TypeScript should work", I'd like to draw a quick summary before offering my own perspective. As of May 2022 things seems to look like this:

- The official stance on the subject seems to be ["we're 100% committed to not rewriting JavaScript code"](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937), effectively prescribing typing `import './foo.js'` whilst referring to `'./foo.ts'` sitting right next to the file you have open — simply because this is how your files will look like when they will be compiled by `tsc`. This conclusion seems to ignore a lot of argumentation stating the behaviour of browsers, Deno, custom Node module resolvers and simply the fact that it doesn't make any sense for relative import to refer to a non-existent resource (I will elaborate on this point a bit later). It also remains a closed conversation, so there is no way for other people to understand whether this stance has changed.
- The issue is raised _a lot_, and keeps being raised and reopened because clearly "something isn't quite right" (i.e. various tools and platforms just don't get along).
- The issue is also a biggie: it comes up every [now and then](https://github.com/denoland/deno/issues/3196#issuecomment-546129170) as a single big blocker towards a greater goal of achieving a more uniform module loading across platforms.
- Interestingly, things are much easier with simple JavaScript modules, because, well, there's only one way of importing a `.js` file that doesn't need to be compiled/transpiled (remember, extensionless imports are out of scope!) — this obviously does not play in TypeScript favour.
- People upset with lack of decisiveness around this seemingly simple issue start to question the general direction of "where all of this is going", so much so that a lot of folks are opting out of TS syntax to stick with JavaScript + TSDoc and/or `.d.ts` (which in my opinion is a considerable step down).
- Circling back to tools:
    - Webpack/ts-loader wants `.ts` imports and won't resolve non-existent `.js` unless `resolve-typescript-plugin` is used
    - Same for Deno: if you import a relative `.js`, it expects the actual file to exist relatively to the file you're currently editing (which makes total sense)
    - Same is true for virtually any other platform that wants to avoid probing for extensions (e.g. Node + custom resolver that tries to not do a lot of potentially expensive calls to a remote registry).
    - TSC and VSCode will both complain on relative `.ts` imports saying "An import path cannot end with a '.ts' extension". 

Now _as a human_ the way I have always imagined relative paths work is: you see something like `./foo/bar` you think "ok, where am I right now?" — depending on the answer to that question you'll know how to resolve that relative path to an absolute one.

Specifically:

- if I'm in my editor, I have a _file_ open. If it's a TS file at `src/main/foo.ts` and it says `import './bar.js'`, then my instinct would be too look for `src/main/bar.js` — but never at `out/main/bar.js`.
- if I'm looking at a script loaded from `https://example.com/coo/paa/foo.js" and it says `import '../bar.js'` inside, then my instinct is to look at `https://example.com/coo/bar.js`; if it says `import './baz.ts'`, then I'm looking at `https://example.com/coo/paa/baz.ts`.
- The two examples above are pretty in-line with how Browsers and Deno work.

Finally, in my opinion, saying `import './foo.js'` doesn't make too much sense not only because the actual file doesn't exist relatively to the source I'm editing, but also because it says `.js` — but implies also importing `.d.ts` (at least implicitly) for the type checking to work.

So given all the above & beyond, I think it's no longer fair nor feasible to simply shrug off the issue with a "wontfix because of a design goal". Big issues like these may warrant a change in the design goals if it's for a greater good.

**Comment by frank-dspeed** [comment]
@inca your observations are all correct but i tracked that also down even more deep.
- A function should only do 1 thing and the resolve stuff always does many splitting the commands into more meaning full once would clear the dust once and for all and would even allow to adopt more fast in other engines
  - importJson, importScript, import /* the ecmascript module one with fetch */, 
- taking the same syntax and produce diffrent results is always headache.
- using the .js extension + JSDOC Annotations or with generated .d.ts is the only useable way avoiding .ts extension as it leads to mixup headache is always speeding up coding. 
- avoid mixed aliases that return dynamic results and replace them with clear single meaning entrypoints that shim the environment.
- avoid environment detection and find other ways to make clear your running in the correct environment let environment detection never be part of your main code.

## conclusion 
as .js + generated or handwritten .d.ts is the only usabele formart to author and ship without quirks the whole issue should be closed wont-fix is the right strategie as .ts needs to die anyway.

**Comment by Conaclos** [comment]
@[frank-dspeed](https://github.com/frank-dspeed)
Unfortunately it is currently impossible to correctly type a file using declaration files or JSDocs.

They nicely allow to type-check code boundaries. However they fail to type-check local code.

Indeed if you use the strictest TypeScript settings (strictNullCheck, ...), then it always ends with a code that does not pass the type-checker. Workarounds are still sub-optimal or there are simply non-existent.

Moreover, JSDocs are less expressive than TypeScript. It makes things more complex. There is some stuff I still do not know how to do with JSDocs (for instance import a type from another file, declare a new type).

Unfortunately, declaration files are not an option here: when you type-check a file it odes not use its declaration file to infer its function/constant types. I opened an issue about that #44946.

The best hope could be the recent [TC89 proposal](https://github.com/tc39/proposal-type-annotations) to natively support type annotations in JavaScript. However, we need to wait several years...

Another solution could be to switch to Flow that has a strong support for types in comments and allow type annotations in js files.

**Comment by cefn** [comment]
@arendjr wrote

> I will happily join an angry mob asking for the import "./foo.ts" → import "./foo.js"

As a villager in the JS and TS ecosystem, my own fork is impossibly ambitious, so I'm left only with my pitchfork :)

This was a frustrating error to encounter as part of an ESM migration in which I had actually got nothing wrong (by way of "type", "module", "moduleResolution" and ts-node ESM loader, each of which was hard-won knowledge with [little or no meaningful feedback](https://github.com/cefn/zx-tsnode-repro/issues/1)). Then the first time I used the configuration 'in anger', (with a relative import), I found a [new inexplicable error](https://github.com/cefn/zx-tsnode-repro/pull/5)  I couldn't expect colleagues to go through this to get perfectly functional code to run. This endangers the adoption of the tools.

I interpreted the inability to resolve files (even when the path was exactly right) as something wrong in my config as I could see the file, and had always imported .ts files without suffix before. This meant another whole round of experimenting with flags and config in 4 places to soothe the ESM dependency monster. I tried an explicit .ts which seemed reasonable to clarify the path and it was rejected. `.js` wouldn't have occurred to me in a million years.

**Comment by merrywhether** [comment]
With the rise of `babel-ts`, `esbuild`, and `swc` providing a variety of options for people looking to work with TypeScript, is it necessary for `tsc` the type-checker (as the avatar of TypeScript the language specification) to stay perfectly in sync with `tsc` the compiler/transpiler? Having the type system be a superset of the compiler capabilities would free up language design from generalizable implementation complexity. Language design would still need to _consider_ complexity, but pushing some things out into community-only implementations would free any single solution from having to accommodate _every_ usage pattern and could instead have its own opinions about which combinations are supported. That means that `tsc` transpilation can continue to focus on ease-of-use/-configuration and hew towards principles like "no rewriting JavaScript" while something like `babel` transpilation can embrace its wide support of happily rewriting lots of JavaScript. I think web devs are used to the idea of "exotic features require more config" and manage those trade-offs all the time, so this wouldn't be a major departure.

In this case, TypeScript would be freed up to add the flag while the "easy" route of using only `tsc` for transpilation could continue its design arc. These extensions could pass-through `tsc` unchanged and using them would require replacing or augmenting `tsc` in a project's build pipeline, just like people have to do if they opt into `styled-components` or `relay` or any of the many other transform-requiring libraries out there. And then projects can use finely-/singly-tuned solutions that enable `./foo -> ./foo.js` or `./foo.ts -> ./foo.js` or whatever else crops up if they feel the choice is worth the complexity, and everyone is happily building apps the way they want.

This would obviously be a big change of stance for TypeScript, but it would enable a path forward for things like this. And making this divergence explicit should quell the "angry mobs" who you can just point to another transpilation solution if they want to get alternative transpilation outcomes.

**Comment by frank-dspeed** [comment]
@merrywhether there is something new upcoming i hope i get it finished till christmas but i am Implementing the Whole Typescript Language at present on Truffle as also a Special GraalJS Branch that includes https://github.com/tc39/proposal-type-annotations

This will allow you to base on that for your own Language Implementations. 

Graal stands for „General Recursive Applicative and Algorithmic Language“. There is something we call GraalVM it is able to implement any „General Recursive Applicative and Algorithmic Language“. via the so called Truffle Framework. 

So You can see it as Coding Language Development framework that is fully debuggable via devtools as typescript is.

**Comment by andrewbranch** [comment]
Currently my biggest hesitation is in what to do in checking .js files.

```ts
// @Filename: a.ts
export const a: string = "hello";

// @Filename: b.js <-- ⚠️ JS, not TS!
import { a } from "./a.ts";
```

Lots of people check their JS with `tsc --noEmit` and run it directly in Node or the browser, so it sort of seems like a bad default to let `--noEmit` alone enable this behavior. But on the other hand, presumably every bundler is perfectly happy with this. (Also I assume this is fine in Deno?) Currently leaning towards “this is fine,” since if you’re tempted to write this in a JS file, you have a mix of JS and TS files and either

- you know exactly what you’re doing and it’s fine, or
- you _are_ using `tsc` for emit, and you’ll get the same error we issue today.

But it is a bit of a weird effect.
- _not_

**Comment by frank-dspeed** [comment]
@andrewbranch out of bundler view in case of rollup importing "./a.ts";  is fine it will work without errors with emit with noEmit it will emit nothing so it will not work. At last in case of rollup.

**Comment by jespertheend** [comment]
> Lots of people check their JS with tsc --noEmit and run it directly in Node or the browser, so it sort of seems like a bad default to let --noEmit alone enable this behavior.

I'm exclusively using .js files with `--noEmit` with a few exceptions where I use `types.d.ts` or `types.ts`, for when writing types using JSDoc becomes too tedious. They don't contain any runtime code and are imported using JSDoc tags like
```js
/** @type {import("./types.js").MyComplexType} */
const x = "hello";
```

This is not an issue when running in the browser (without a build step) because it doesn't import these files. So it doesn't matter if they are imported with `.js`, `.ts` or no extension.

This method does fail when using Deno for running tests however, since it expects `types.ts` to actually exist. So changing the behavior based on whether `--noEmit` is used would solve this. The only two issues with this approach I can see with this is that:

- Importing `.js` files that are actually `.ts` or `.d.ts` files (using regular imports, not JSDoc) wouldn't cause any TypeScript errors. Even though Deno would definitely fail when running this kind of code. But this is already an issue at the moment, and you can easily find out that this is the case by just running your application (unless you're using dynamic imports), or by running your bundler if you have one.
- The language server doesn't know if `--noEmit` is being used. So errors will still show up in your text editor. In that sense I think an option in `jsconfig.json`/`tsconfig.json` would be preferable.

**Comment by Conaclos** [comment]
> * The language server doesn't know if `--noEmit` is being used. So errors will still show up in your text editor. In that sense I think an option in `jsconfig.json`/`tsconfig.json` would be preferable.

If you don't use TSC to transpile your source files, then you can use the `noEMit` option in `tsconfig.json`. The language server relies on `tsconfig.json`.

**Comment by jespertheend** [comment]
Ah cool, I thought noEmit was only available as a command line flag.

**Comment by frank-dspeed** [comment]
@jespertheend as rule of thumb all cli flags have a eq setting in the tsconfig

**Comment by andrewbranch** [comment]
I want to give an update of my investigation into this here to outline some decisions we’ll have to make.

## Background

There are two separate systems involved when you write `import {} from "./foo.ts"`: module resolution and type checking. Module resolution is what takes the input filename and the string `"./foo.ts"` and queries the disk to find what file you’re talking about. That result is saved and used by the type checker later. The type checker is the one that today issues the error that says you’re not allowed to use `.ts` in module specifiers.

## Module resolution today

My assumption, until I started prototyping support for this feature, was that module resolution to `.ts` files given a `.ts` specifier was succeeding, and the type checker was subsequently issuing an error not because resolution failed, but because by the logic of our emit, the code would _later_ fail at runtime. If this were true, it would mean we could implement this without breaking changes—we could simply stop issuing the error under the right circumstances, and everything else would Just Work™. This assumption was incorrect: because module resolution knows that `.ts` extensions aren’t supposed to be valid specifiers, it skips looking up a file by that exact name and (depending on the `moduleResolution` compiler setting) moves straight on to fallbacks. E.g., for `./foo.ts` in CJS Node resolution, we will look for `foo.ts.ts`, `foo.ts.d.ts`, `foo.ts/index.ts`, etc., without ever looking for `foo.ts` itself.

In fact, this behavior is not specific to `.ts` files. Consider the file structure (and I’m going to use `.js` files here to set up an argument):

```
foo.txt/
├─ index.js
foo.txt
main.js
```

If we `require("./foo.txt")` from `main.js`, what should happen? In Node, the request resolves to the `foo.txt` text file, and it’s assumed to be JavaScript—Node doesn’t care about file extensions. If the contents don’t parse as valid JS, it will crash. But in TypeScript, because we don’t recognize `.txt` as a file extension that we will read, we don’t even check if it exists; we simply move on to fallbacks and eventually resolve to `foo.txt/index.js`.

## Why this is a problem

To be clear, I am _not_ arguing that TypeScript should load and read files regardless of their file extension. However, I do think that it could be considered a long-standing bug of some of our existing module resolution strategies that we don’t even probe for _existence_ of these files, since the above example shows TypeScript successfully resolving to some file, while the module resolver it intends to model resolves to a different one and probably crashes.

This behavior must change to allow resolution to `.ts` files. However, it is untenable to change the behavior of module resolution based on `--noEmit`. If `--noEmit` is to be the only criteria for allowing resolution to `.ts` files, it means module resolution must _always_ succeed in finding `.ts` files by explicit extension, even if an error is issued later.

## The decision

The primary decision we have to make is under what settings we fix this bug, which directly affects where `.ts` extension resolution can be supported.

### Option 1: Fix it everywhere (with flag to opt out)

This would be a breaking change—and the kind of breaking change we _really_ don’t like to make: one that affects the super old, super stable `node` module resolution mode—but, via the example I gave earlier, this _is_ a hole in our current ability to catch actual problems, so fixing it _does_ have some benefits aside from enabling new features.

### Option 2: Add a flag specifically for allowing `.ts` resolution, fix it under that

Such a flag would imply/enforce `--noEmit`, and there is some documentation/discoverability benefit to making the feature explicit and opt-in.

### Option 3: Do not treat `.ts` resolution as its own feature; lump it into new `moduleResolution` mode(s)

If we take a step back, the motivation for this feature is that there are module resolvers out there—namely bundlers and Deno—that process TS inputs and allow (or even require) them to reference each other with `.ts` extensions. Fundamentally, these module resolvers ought to be represented by a `moduleResolution` setting made for them that includes `.ts` resolution as well as the precise set of other features they support. The idea that you could continue to use `--moduleResolution node` (or any other existing mode) but turn on some additional flag and/or set `--noEmit` and have that change resolution is a band-aid over the fact that we don’t have full `moduleResolution` setting appropriate for these resolvers.

To me, reasoning through this argument makes every other approach feel like a bit of a hack. The problem with this approach, though, is it would mean we have a lot of work to do before we can ship anything, which feels very disappointing given that just a couple days ago I thought module resolution was already working in a way where we could slip this feature in with no breaks and no new flags and it would ease a lot of the pain that currently exists with using bundlers and Deno. (Specifically, the thing that needs to be decided if we go down this route is how to deal with the sheer volume of module resolvers that exist in the wild today and the fact that many of them are infinitely configurable—it’s not at all clear where to draw the line of what’s supported and what’s configurable and what’s not.)

## Conclusion (sort of)

I will continue to get feedback from the team and try to push us toward some kind of consensus. I think getting our support for bundlers and other widely-used non-Node module resolvers to a usable state (especially respecting conditional exports: see recent conversation at #33079) is super important—this focus on module specifier extensions is really just a small facet of a bigger problem.

<details><summary><strong>A postscript of reflection on tradeoffs, time horizons, and a personal bugbear</strong></summary>

I also hope this sheds some light on the fact that these kinds of changes can be dramatically more complex than they appear. @arendjr’s comment has been nagging at me for over a month:

> This seems to be a classic defeatist argument. Just because you cannot solve _all_ problems should not be an argument to not solve _any_.

I have just shown how a bug or design decision (we may never know) written the better part of a decade ago is boxing us out of being able to ship a simple fix for this issue. Instead, we have to weigh breaking changes against increasing the complexity of an already complex configuration space. There are so many things that Present TS Team would have done differently from Past TS Team. Trying to give Future TS Team bandwidth to react to unpredictable things that happen between now and then is important for a project of our time horizons. I think it’s a fair criticism to say that we’ve been very biased toward delay when weighing these concerns when it comes to modern module stuff specifically. But I want to push back against the idea that I’m just sitting on a 90% solution and not giving it to you because the remaining 10% is keeping me up at night. No. The majority of things I’ve shipped here have been compromises, heuristics, and best-effort attempts to synthesize a veneer of order over the messy confluence of systems that is the JS ecosystem. The culmination of this work will surely fit that description too. But given how interconnected all these module-related concerns are, and how fragmented the ecosystem is, it’s worth trying to get this as right as we can get it.
</details>

**Comment by johnnyreilly** [comment]
This is quite besides the point @andrewbranch , but I wanted to say that I've often been impressed with the articulate way you write up your thoughts and reasonings around things. It's an excellent habit that I try to emulate.  Well done sir! ❤️🌻

**Comment by inca** [comment]
Sincerely appreciating all the arrrghumentaion above ❤️

Still I think .ts + noEmit option is even more questionable, and am frankly quite surprised that it's even on the table.

I cannot possibly come up with the arguments as strong and as accurately articulated as @andrewbranch made, and I can totally understand the amount of planes of complexity the problem resides in, as well as compatibility, future-proofing and related concerns. Almost all software I've ever seen in my life tend to evolve into something I tend to compare to [overdetermined system](https://en.wikipedia.org/wiki/Overdetermined_system) whereby the "clean" solutions are no longer possible due to a sheer amount of constraints. So what happens is that the majority of decisions are made to be "least worst" instead of "right". They further reinforces the cycle of software obsolescence (as is often depicted by "software failure curves" and similar metaphors).

This is why I think it's equally (arguably, even more than equally) important to try and not to loose the sight of the "top of the pyramid", the "why", the whole premise of TypeScript existence. The way I understand it, TypeScript's root premise (aside from ruling the World) is to improve the DX of JavaScript. I mean, if it's not about the DX, then even types make no sense (computers don't really care about the types ultimately, only humans do). And of course DX is much more than just the types: things like `tsc`, `tsconfig.json`, VSCode and friends all play a crucial role in forming this DX entity. So it's only reasonable to expect that the decisions need to be aligned with supporting the well-being of this entity, one way or another. 

Now, back to .ts + noEmit. So previously we had a problem that bundlers would expect the `.ts` extension (because it just makes sense, the file is right here, take it) whilst `tsc` would expect the `.js` extension (because TS refuses to modify the emitted imports). And now we have a whole different problem: `tsc` does resolve `.ts` but can no longer emit.

I don't know about the others, but we only use `tsc` to compile Node.js codebases (which still needs the JS files to be emitted, and if `.ts` is specified in source imports, they have to be re-written into `.js`) — and I can't think of any other use case from top of my head. I mean, Deno doesn't need `tsc` and with bundlers you don't use `tsc` directly. Node.js cannot run `.ts`, so it still needs emit, which makes .ts + noEmit combination almost completely useless. What am I missing here? What exactly did we solve? More importantly, how did the DX improve?

Once again, this is not an attempt to criticize the decision making, or to diminish the value of the amazing work you are doing. It's an attempt to see a little bit above the current horizon. JS ecosystem is already in a bad shape — don't add more to it. For example, if backwards compatibility with "super old, super stable" CommonJS is the reason why those who try writing ESM full-stop still struggle, then... Just drop the support for emitting CommonJS! Seriously. Release a major version, or put it behind a flag, name it differently — not important in the grand scheme of things. In this new mode, allow (enforce?) `.ts` extension in static imports an emit ESM with imports modified to `.js`. TSC doesn't have to know how to emit the existing non-ESM libraries — it only needs to know how to resolve the `.d.ts` files, so the libraries will just work in Node.js runtime. Also gives you guys an opportunity to clean up some of the other things that nobody uses (e.g. ability to emit everything into a single file). 

Of course, this is just an example, but let's consider the UX part of such change.

1. _Initially_ makes a lot of people unhappy/angry. They expected that their codebase emitted as CommonJS will just continue to work with the major bump of TypeScript, but it doesn't.
2. However, they read about the change in the blog and quickly realise that all they have to do is to just re-write the imports a little (add extensions and `index.ts` here and there). A small price to pay to remain future-proof.
3. Those who cannot afford the change can stay on v4 until they can. Not ideal, but not the end of the World.
4. Now the other camp, the ESM people, can finally sleep a little better at night, knowing that their backends and frontends now use the exact same import syntax and module resolution strategy, the code emitted by `tsc` works perfectly both in Node.js and even natively in browsers with `type="module"`. Also they don't have to make separate editor configs and they can finally remove the hacks like `ResolveTypeScriptPlugin` from their huge `webpack.config.json`s.
5. Existing Node.js libraries don't have to adapt instantly, because the change affects the source code — but the output remains the same.
6. Last but not least, it contributes an important milestone into deprecating CommonJS and Node.js module resolution. Definitely has a longer term net positive on the entire ecosystem.

Once again, I am aware that I might be over-simplifying and that the real life is more difficult. On the other hand, change the perspective a little — and everything becomes relatively simple again: you either contributing to a longer term benefits to the ecosystem by reducing the number of options, or you're contributing more complexity to deal with later on. How to see the longer term benefits? Nobody knows, I'm afraid, but time has a way of telling. 🤷‍♂️

**Comment by frank-dspeed** [comment]
just my 5cent we should put it if it gets putted some where into a own resolveMode 

while i would suggest to drop the whole resolve stuff in favor of classic as classic is the only mode that makes sense to engines and bundlers classic simply needs to be aware of js mjs cjs thats it every other method is doomed to cause convergences.

my standard fix for every single dependencie of any of my projects is it to transpile the original source code when authored in .ts to match the classic resolve pattern that saves me latter when i am in dev from any convergences.

finally i use a bundler for my project and bundle also the d.ts files correct as it is easy to work with them as i can always expect classic resolve to work.

**Comment by Jack-Works** [comment]
Is it possible to do the normal TS to JS compile, but the output folder also emitting `.ts` as the result extension?

> src/index.ts

```ts
import { x, type y } './a.ts'
```

Output will be

> dist/index.ts

```ts
import { x } from './a.ts'
```

**Comment by frank-dspeed** [comment]
@Jack-Works that is possible but that will lead to some other edgecases depending on where the output happens not every one uses a outDir or declarationDir.

**Comment by tukusejssirs** [comment]
> [@andrewbranch]
>
> foo.txt/
> ├─ index.js
> foo.txt
> main.js

Just an OT note: it is not possible to create two files (or folders) with the same name in a single folder.

**Comment by cefn** [comment]
I accept this is tooling bug (given the current implementation) but I found Intellisense was struggling with imports from `.js` extensions in VSCode (I believe it wasn't able to identify any such file because it doesn't exist). 

Needing a special case mechanism in IDEs to treat suffixed .js as mapped to .ts is another data point for having Typescript source files pointing to other actual source files, whatever the eventual structure that's needed in the transpiled code.

**Comment by Pyrolistical** [comment]
One way to move forward without breaking everybody's existing code is to deprecate `tsc` and fork it into 2 commands:

1. typescript type-checker `tstc`: only does type checking
2. typescript bundler `tsb`: bundles js/ts files into one or many esm files

Both commands are esm compliant and require .ts extension.

Keep shipping `tsc` with existing behaviour but with a scary deprecated warning that `tsc` will be removed next version. After it is removed leave a placeholder `tsc` that links to an article on how to migrate to `tstc`/`tsb`.

Why break into 2 commands? Because that's what the community already did. `tsc` was so slow at emitting js from ts that babel-ts, esbuild and swc were created. VS Code lang server only needs `tstc`.

This would also allow the typescript team to keep their position on not rewriting javascript as part of "emitting" as now `tsb` is doing _bundling_ which its job is to rewrite all the files.

**Comment by andrewbranch** [comment]
Why on earth would people want to use us as a bundler when esbuild and swc are fast, popular, and high-quality? We are more than happy for people to use these tools. The way to “move forward without breaking everybody’s code” is to add a new setting to `--moduleResolution`. But sure, technically, reimagining the scope of the entire project, deprecating everything that existed previously, leaving millions of users who do normal builds with tsc behind on that deprecated path, and writing a new bundler from the ground up that will be vastly slower than its competition is also one possible solution. 😄

**Comment by RyanCavanaugh** [comment]
To put it another way, we have finite time. We can either:
 * Add new type system features, better editor features, improve error messages, just generally make the experience of using types in JavaScript better
 * Re-implement functionality already available in widely-used, well-liked, TS-friendly bundlers

**Comment by ritschwumm** [comment]
@RyanCavanaugh it seems our experiences with typescript differ.

i'm quite happy (actually, very happy) with the type system, and the editor features.

what i'm not happy with is the UX of tsc. in 30 years of professional software development, 
i've never before spent so much time on understanding how all the options and variants
of module resolution and compiler output interact with the already complicated node
package ecosystem.

i think i can assume that all this is not a problem for you at all: you probably have
_seen_ all these options being built and evolve, and can remember their purpose, and what 
they do.

for someone who only _uses_ tsc, as opposed to actively participating in its development,
the picture is quite different. so please, please at least acknowledge that even when you
do no see any problems with the status quo, other people might. imho, typescript could be
a lot friendlier to "the rest of us".

**Comment by Pyrolistical** [comment]
@RyanCavanaugh I agree, the TypeScript team should focus on new features and not implementing a bundler.

Given that, maybe consider deprecating emitting from `tsc`? We already have 3 independent projects for emitting/bundling from the community. And probably even more if https://github.com/tc39/proposal-type-annotations lands.

TypeScript has won our hearts and minds. It is being embraced by the community. Perhaps it is time to hand over parts of TypeScript to the community as it is now mature enough to self-sustain.

**Comment by rauschma** [comment]
The use case I’m seeing is supporting TypeScript on Node.js – e.g. via loader hooks: https://github.com/lukeed/tsm

Maybe that functionality will eventually even be built into Node.js. If the filename extensions `.ts`, `.tsx` etc. were allowed, static type-checking would work at runtime. Node.js would simply strip away the type information (plus maybe a few simple transpilations).

**Comment by inca** [comment]
Ugh, tsc + noEmit, bundlers, revolutions. Just tell me, what do I need to use to compile a bunch of TypeScript files from `src` into a bunch of JavaScript files at `out`? So that things like Node.js could execute stuff from `out` and things like Deno can execute stuff from `src`.

**Comment by paralin** [comment]
@inca `esbuild ./src/main.ts --outdir ./out --bundle`

**Comment by RyanCavanaugh** [comment]
Deno and NodeJS are different runtimes and there's *in general* not a way to write a program for one runtime and have it happen to execute against a different runtime. From the TS perspective, switching runtime targets is the job of a bundler.

**Comment by paralin** [comment]
@donmccurdy I see what you're saying. Being able to import the .ts suffix in the /pre-bundled/ code would definitely disambiguate a lot of situations. For example, when using a file like "thing.pb.ts" importing "thing.pb" doesn't work and "thing.pb.js" works, but confuses a lot of developers to the point that they file an issue complaining, and doesn't work in typescript < 4.7.x. 

@RyanCavanaugh there's really no reason why deno and node.js shouldn't be able to execute the same code without changes, besides the limits put in place (for dubious reasons) by the developers of the compiler.

**Comment by RyanCavanaugh** [comment]
I agree that deno and node.js should support identical module resolution schemes

**Comment by donmccurdy** [comment]
> From the TS perspective, switching runtime targets is the job of a bundler.

I'm OK with this. But at the same time, resolving imports in executable code is the job of the runtime. The Deno runtime has made a reasonable choice for their needs in resolving explicit `.ts` paths. TS is making assumptions about what type of modules the runtime might be able to load. I can import from `image.png` without error, configuring TS to trust that something (in the runtime or the bundler) will resolve that import. But TS does not extend that same benefit of the doubt to imports with `.ts` extensions. I'm happy to use a third-party bundler to compile my code to JS, but it's frustrating that we need a VSCode extension to hide errors about `.ts` extensions, when Deno is a significant runtime environment for the JS/TS ecosystem.

**Comment by andrewbranch** [comment]
The discourse today make me think that maybe some people interpreted my earlier message and/or the state of the labels to mean that we have no idea how to address this issue or even that we still need to be convinced that it deserves action. I am not milestoning it yet because I don’t yet feel confident enough to make a public prediction about when I can land something, but I have a good grasp on the success criteria and constraints, and I’m gathering data and discussing a few concrete approaches with people on the team. Discussion is always welcome, but it would make my work easier if it becomes more focused as we narrow in on a course of action. This is not a catchall issue about the UX of tsc or the state of modules in JS, nor is it at a stage where ideas like “deprecate emit” are going to be considered, nor will expressing frustration at the current state of TypeScript make a solution appear any sooner.

The scope of this issue is that there are many module resolvers built into bundlers and non-Node runtimes where importing files by `.ts` extensions makes sense, and TypeScript does not have module resolution modes or options to support this. It should. Related, but _out_ of scope, is the idea that you should always be able to write and check a TypeScript program targeting an arbitrary set of module resolvers.

**Comment by rauschma** [comment]
The most important use cases for the filename extension `.ts` that I can think of (especially long term) are: Deno, Bun, [tsm](https://github.com/lukeed/tsm). They all:

* use relative module specifiers
* to refer to ESM modules.

I don’t think you need `.ts` for bare specifiers and full URIs – but I may be missing something.

**Comment by Jamesernator** [comment]
> I don’t think you need `.ts` for bare specifiers and full URIs – but I may be missing something.

Deno does use `.ts` in full URLs for fetching from on the web (including its [own standard library](https://deno.land/std@0.148.0/path)).

**Comment by izaakschroeder** [comment]
For what it's worth this crude hack (+ `yarn patch typescript`) makes `tsc` accept `.ts` imports:

```diff
--- a/lib/tsc.js
+++ b/lib/tsc.js
@@ -34960,6 +34960,7 @@ var ts;
     }
     var Extensions;
     (function (Extensions) {
+        Extensions[Extensions["None"] = 6] = "None";
         Extensions[Extensions["TypeScript"] = 0] = "TypeScript";
         Extensions[Extensions["JavaScript"] = 1] = "JavaScript";
         Extensions[Extensions["Json"] = 2] = "Json";
@@ -35823,7 +35824,7 @@ var ts;
             requestContainingDirectory: containingDirectory,
             reportDiagnostic: function (diag) { return void diagnostics.push(diag); },
         };
-        var result = ts.forEach(extensions, function (ext) { return tryResolve(ext); });
+        var result = ts.forEach([Extensions.None].concat(extensions), function (ext) { return tryResolve(ext); });
         return createResolvedModuleWithFailedLookupLocations((_a = result === null || result === void 0 ? void 0 : result.value) === null || _a === void 0 ? void 0 : _a.resolved, (_b = result === null || result === void 0 ? void 0 : result.value) === null || _b === void 0 ? void 0 : _b.isExternalLibraryImport, failedLookupLocations, affectingLocations, diagnostics, state.resultFromCache);
         function tryResolve(extensions) {
             var loader = function (extensions, candidate, onlyRecordFailures, state) { return nodeLoadModuleByRelativeName(extensions, candidate, onlyRecordFailures, state, true); };
@@ -35948,6 +35949,12 @@ var ts;
             var extension = extensionLess ? candidate.substring(extensionLess.length) : "";
             return (extensionLess === undefined && extensions === Extensions.Json) ? undefined : tryAddingExtensions(extensionLess || candidate, extensions, extension, onlyRecordFailures, state);
         }
+        if (extensions === Extensions.None) {
+           var path = tryFile(candidate, onlyRecordFailures, state);
+           var extensionless = ts.removeFileExtension(candidate);
+           var extension = candidate.substring(extensionless.length);
+           return path === undefined ? undefined : { path: path, ext: extension };
+        }
         if (!(state.features & NodeResolutionFeatures.EsmMode)) {
             var resolvedByAddingExtension = tryAddingExtensions(candidate, extensions, "", onlyRecordFailures, state);
             if (resolvedByAddingExtension) {
@@ -36226,6 +36235,8 @@ var ts;
         var packageFile;
         if (jsonContent) {
             switch (extensions) {
+                case Extensions.None:
+                    break;
                 case Extensions.JavaScript:
                 case Extensions.Json:
                 case Extensions.TsOnly:
@@ -42572,16 +42583,7 @@ var ts;
                     var moduleResolutionKind = ts.getEmitModuleResolutionKind(compilerOptions);
                     var resolutionIsNode16OrNext = moduleResolutionKind === ts.ModuleResolutionKind.Node16 ||
                         moduleResolutionKind === ts.ModuleResolutionKind.NodeNext;
-                    if (tsExtension) {
-                        var diag = ts.Diagnostics.An_import_path_cannot_end_with_a_0_extension_Consider_importing_1_instead;
-                        var importSourceWithoutExtension = ts.removeExtension(moduleReference, tsExtension);
-                        var replacedImportSource = importSourceWithoutExtension;
-                        if (moduleKind >= ts.ModuleKind.ES2015) {
-                            replacedImportSource += tsExtension === ".mts" ? ".mjs" : tsExtension === ".cts" ? ".cjs" : ".js";
-                        }
-                        error(errorNode, diag, tsExtension, replacedImportSource);
-                    }
-                    else if (!compilerOptions.resolveJsonModule &&
+                    if (!compilerOptions.resolveJsonModule &&
                         ts.fileExtensionIs(moduleReference, ".json") &&
                         ts.getEmitModuleResolutionKind(compilerOptions) !== ts.ModuleResolutionKind.Classic &&
                         ts.hasJsonModuleEmitEnabled(compilerOptions)) {
```

**Comment by frank-dspeed** [comment]
the main problem is resolve it self:
some will prefer to resolve .js => .ts and some will want .ts => .js  that depends total on usage so there is no good default for modules without any specifier.

just my 5 cent. but i would suggest droping bareSpecifierSupport for typescript as typescript is filebased only . that would make the most sense at present. 

while there also is a option to use bareSpecifiers only but that would need engine support the only way to get the engine to take bare specifiers would be a import map via package.json or import maps nativ. 

so again most elegant option that all involved support is file: protocol and so file specifiers. 

```
src/files/a.js wants to import b from 'b' with node module resolution both will stat locking for b in ./node_modules/b or /node_modules/b or package.json map for node and tsconfig alias map for typescript
```

the main problem is there could be many places where a user defines package mappings 

The only way to let typescript do its job would be to hook it into the specifier resolver of the host runtime or even a custom one 

i propose implementing a resolve hook into typescript that is configure able via tsconfig.json



how to we get b to be in node_modules?

**Comment by ctjlewis** [comment]
Presumably these imports will not be rewritten, and so we will still not be able to build valid ESM programs without doing the "fake `.js`" hack that this team somehow thought was a sustainable solution.

Refusal to support specifier rewrites the biggest obstacle to ESM adoption at the moment. Not only is it not prioritized, but we're going deeper down the rabbit hole, where now you can have `.ts` specifiers that *also* won't resolve at runtime!

https://github.com/microsoft/TypeScript/issues/42151

**Comment by paralin** [comment]
> TS's main goal is to compile to ES and we've seemingly just abandoned that

What? I don't agree. Does anyone run TypeScript without compiling to ES first?

**Comment by ctjlewis** [comment]
> What? I don't agree. Does anyone run TypeScript without compiling to ES first?

You don't agree with what? Yes, many people do run TS without compiling, e.g. Deno, and the .ts import specifier caters to that.

Did you read what I wrote about valid TS-ESM programs not compiling to valid ESM programs?

**Comment by frank-dspeed** [comment]
@ctjlewis It is a superset as it includes partial<esNext> it is a Language a Runtime Interpretet one to be fair. Implemented on Top of the ECMAScript language. 

Technically this is a bad coded code generator with some syntax checking. We call that often Domain Specific Language. the Domain of Typescript is to produce some what around Javascript EcmaScript running on a subset of the implementing environments if when they are able to express the environment for the code generator (TS) The Most big goals that Typescript did Archive was:

- Language Service Protocol 
- The rest was a good try but we see it did fail like flow and any other linter.

But there is also light on the Horizon.

The Concepts are good and good ECMAScript linters and Toolchains can speed up development i guess there is no one to blame This is open source and no one did come up with some better concepts until now or they where at last not able to implement them. 

For me typescript was something of the most inspiring things to see because of the way it was Implemented it Opened my Mind for the Idea of Polyglot Runtimes implemented on top of existing Runtime Platforms . I was never thinking that i would be able to Implement such stuff but Typescript did prove that. I was always thinking i would need to code a Whole engine from scratch to Implement a usefull Language.

**Comment by RyanCavanaugh** [comment]
@frank-dspeed please keep discussion on-topic

**Comment by ctjlewis** [comment]
> So I don't see why adding another flag, "suffixESImportsOrWhatever" that transforms imports in a slightly different way would be too much of a stretch to imagine.

Exactly correct. There is no obstacle to adding non-breaking, opt-in support for this. 

Noted in the issue, someone actually added this flag and sent in a PR, but it was summarily closed. It is a striking example of organizational deficiency in software. I suspect a single decision-maker or a small group of them are the only obstacles to un-breaking the compiler in this respect.

**Comment by thetutlage** [comment]
Does this mean that a file using `.ts` extension in a relative import will retain the same extension after it gets compiled to JavaScript?

**Comment by frank-dspeed** [comment]
@thetutlage nope it means that there is a .ts even allowed before compiling it to js. Today Typescript is not aware of file extensions in its core it treats everything as Module or Script that are the 2 types that Typescript understands and a Module has no concept of a fileextension as a Module is not related to a file. 

so internal: MyModule.ts is the Same as MyModule.js as Both are simply MyModule hope that makes sense.

and this talks about hornoring the extension which is not really possible as i pointed out above :) so it is a looping problem

**Comment by andrewbranch** [comment]
@thetutlage yes, so the plan is to only allow `.ts` extensions in relative imports when `--noEmit` is set.

**Comment by MicahZoltu** [comment]
Why would you refuse to emit with a `.ts` extension?  Emitting *no* extension is already broken (won't run in any environment other than legacy NodeJS), so emitting `.ts` wouldn't make things *worse*.

Allowing emit of an untransformed `.ts` would at least make it easier for external tooling to transform paths, as it can just look for any file with `.ts` suffix and transform it rather than trying to guess whether a suffixless import was a TS file or an actual file without a suffix.

On top of that, if I am writing code that is meant to run in either deno or ts-node or ES Modules, I would want to write it with `.ts` extensions (which IMO is the correct way to write imports of TS files), and then have a post-compile transformer script that fixes the broken emitted code, rather than having to write code that doesn't work in deno or ts-node and having to require some special integration in those tools that fixes the broken un-emitted code).

**Comment by frank-dspeed** [comment]
@MicahZoltu this is correct and is a sympthom of the root cause that .js is not able to check it self when it references existing d.ts files the root cause is a circle of deependencys it is like with RHEL we got the same high level situation here.

So we need to apply the same solutions here:
FEDORA is the Experimental RHEL release CENTOS gets drived from RHEL and CENTOS is needed to get Packages Into Fedora and RHEL . so you need to stream back and for from REHL to CENTOS thats why they made now CentOS Stream. 

Typescript has the same problem .ts is the Experimental .js Release which is then the link to .d.ts and we need to stream back and for between .js .d.ts to get .ts again.

The real solution would be tackling all the JS .d.ts isssues that would enable direct authoring in .js as .js is the only run able real format not .ts

while there exists workarounds to make .ts run it would be the right thing to execute the ECMAScript Annotation proposal that would be done via finishing the babel integration and then update the proposal as soon as the babel work is done i will push my changes about that proposal into the Oracle Java Stack (GraalJS) this will enable the propsal then in all JDK Versions and all JS / ECMAScript integrations on the Java Platform so the proposal is then use able already by Millions out of the box.

Next step create a small typescript tool that only detects none annotation based typescript features integrate that into babel and typescript it self to give usefull warnings when such typescript features are in the code. 

We Should never forget where typescript comes from it is a Facade Impelementation on top of ECMAScript designed to get Compiled away. https://github.com/tc39/proposal-type-annotations

**Comment by josh-hemphill** [comment]
> Why would you refuse to emit with a `.ts` extension? Emitting _no_ extension is already broken (won't run in any environment other than legacy NodeJS), so emitting `.ts` wouldn't make things _worse_.

`--noEmit` doesn't emit without extensions... It instructs tsc to only type check and not emit any code at all. Meaning you can use `.ts` in your IDE/editor and when you actually need to compile you just have to have something change the suffixes before passing to tsc.

**Comment by frank-dspeed** [comment]
@josh-hemphill i guess you have forgotten the secund importent case that typescript gets used for .d.ts generation --noEmit is a project setting and applys when you send the project to tsc that is correct but when you handle the files via diffrent tooling --noEmit does not matter so you can see it simple as indicator that the project gets handled external

**Comment by MicahZoltu** [comment]
> `--noEmit` doesn't emit without extensions... It instructs tsc to only type check and not emit any code at all. Meaning you can use `.ts` in your IDE/editor and when you actually need to compile you just have to have something change the suffixes before passing to tsc (e.g. a loader or some other tool).

I understand what `--no-emit` does, what I don't understand is why you would have the compiler generate an error on emit if an emitted file would have a relative path import that ends in `.ts`.  Given that .ts imports will be resolvable by the compiler prior to the emit phase (after implementing this), you must be going out of your way to have it fail at emit time, and I don't understand why you would do that.

Previously, the argument that @RyanCavanaugh has made is that the TS compiler shouldn't be transforming import paths, so all you would have to do here is continue to follow that reasoning and emit what the user said (e.g., `./path/to/file.ts`).

**Comment by frank-dspeed** [comment]
@MicahZoltu let me explain the internals again in general TS does not Handle any Suffix anything is a Script or Module without a extension JS TS ECMAScript all have no notion of files in general out of Engine view a Runtime does implement that kind of features. 

in general a .ts extension when they should stay internal leads to a lot of confusion about what to resolve and what is authorativ. 

This got internal solved via stripping the file extensions and joining all the files so for example if .js .d.ts exist they are for TS a single Module when .ts exists also this is the single authorativ module 

there exists no internal way at present to tell TS what is Authorativ it has a hardcoded priority .ts => .js .mjs .cjs => d.ts d.mts d.cts

**Comment by MicahZoltu** [comment]
@frank-dspeed Currently, if you do `import 'my.js'` in TS it will emit `import 'my.js'`, and if you do `import 'my'` it will emit `import 'my'`.  This behavior tells me that the compiler/emitter does *not* fully strip the extension because it is able to emit a file with the same extension as you have written in the TS import statement.

For the context of this issue, I don't particularly care about what the compiler is doing internally, I'm only asking that it *doesn't* needlessly error at emit time when the imported file has a `.ts` suffix.

**Comment by frank-dspeed** [comment]
@MicahZoltu what should it do so when it gets a import with .ts should it emit .js ? that would mean path rewriting so we need --noEmit. 

the only way .ts can be perserved is simply do not compile it to js and as soon as that is done you can have .ts

if you then still want to compile to js you need to use external tooling like rollup as this is happy to rewrite that specifiers for you. hope that makes some sense for you i hate the situation overall my self but this is not the only problem and they are all related some how its about module resolution at all.

**Comment by MicahZoltu** [comment]
> @MicahZoltu what should it do so when it gets a import with .ts should it emit .js ? that would mean path rewriting so we need --noEmit.

For the context of this issue, it should emit `import 'my.ts'` since TSC has a policy of never rewriting import paths.  This is the behavior proposed in #38149.

I would *like* for the compiler team to drop that policy and emit `import 'my.js'`, but that is out of scope of this issue.  My latest argument for this can be seen at https://github.com/microsoft/TypeScript/issues/49083#issuecomment-1228160389

**Comment by inca** [comment]
Watching this thread, I have a constant urge to contribute my piece of reasoning, along with once again reminding everyone of greater goals and ideals — but honestly, everything is already said, all the possible reasoning already articulated, more than once. I don't think anyone can contribute anything else — the decision is made. 😢

**Comment by thetutlage** [comment]
I personally think that allowing the `.ts` extension is not a great choice. It just makes the official compiler `tsc` emit code that cannot run the most of the default JavaScript runtimes.

**Comment by nicolo-ribaudo** [comment]
@thetutlage The current proposal is to allow `.ts` when type-cheking, but not when emitting, so `tsc` would not emit code that does not run.

@andrewbranch Mentioned that they might allow emitting it, but only with explicit opt-in (and at that point, if you ask "hey, let me emit invalid code!", then you _know_ that you have to do something to make it runnable).

**Comment by thetutlage** [comment]
@nicolo-ribaudo Thanks for the clarification.

>  The current proposal is to allow .ts when type-cheking, but not when emitting, so tsc would not emit code that does not run

Will tsc fail or will it transform to `.js`?

**Comment by nicolo-ribaudo** [comment]
Yes, because TSC only transform JS-specific parts. As far as the JavaScript spec says, `import "foo.ts"` is valid JavaScript* and thus TS will not transform it.

\* Valid JavaScript that doesn't work with the default behavior of Node.js modules. It works, for example, in browsers or with custom Node.js resolvers.

**Comment by frank-dspeed** [comment]
@nicolo-ribaudo tc39 member here bare specifiers are also part of the spec and the spec is exactly the problem we did a failure as we defined resolve and load to happen inside the Runtime and Not Inside the Engine. Many tc39 members agree in that that specifier lookup in the Runtime was a mistake. the ECMAScript spec defines imports as specifiers and urls so everything is valid. Thats is not the question.

**Comment by MicahZoltu** [comment]
If there is an option to allow emitting of `.ts`, then that satisfies me for the sake of this issue.

**Comment by frank-dspeed** [comment]
i guess this can be bundled now with
- https://nodejs.org/api/cli.html#-ccondition---conditionscondition
- https://nodejs.org/api/packages.html#conditional-exports

this introduces the next problem:
- https://github.com/microsoft/TypeScript/issues/47931

## Relation
Typescript is forced to handle conditional imports that enables conditional import typescript or types 
so when #47931 gets implemented and used .ts emit can always save be done as we 100% know we target typescript it self

**Comment by tracker1** [comment]
I find it funny... TS will read `import foo from "myFile"` go through the trouble of checking for (.tsx, .ts, .jsx, .js) then actually load that first match, and proceed to potentially transform that second file, but actually supporting an explicit file extension, or changing the import statement to match the about to be transformed explicit file output (that is different from input), is too far.

**Comment by frank-dspeed** [comment]
@tracker1 the reason is historical and it is not a bad one at all only this implementation is done bad the reasoning and concepts are in general well it is a bit like with the ECMASpec it self it is easy to write a good spec but implementation and definition do then end up in something diffrent.

The ground concept is the classic resolve algo which is you reference myname.js and it will lookup myname.ts as also myname.d.ts 

that is the base then the assumption comes in that you can not reference .ts files as this would mean to change existing code eg make it point to .js.

when we now want to reference .ts we broke the core concept of TS which is to be a facade over JS that is optional. 

but now as we used all that we come to the conclusion that referencing .ts would be beneficial if we apply external tooling as that saves us from lookups which are error proune.

And now to solve this there are 2 major ways to go and both are none appealing for ts and mirosoft. 

The Both ways to go are:

- Direkt bundler support eg resolve hooks or implementations that fit the needs of external tooling
- Get tc39 type annotation support some how to happen. even if that would be a babel implementation and then throw that into all of the ecosystem that microsoft controls to over vote the externals that are not aware of the big big issues faced by millions.

**Comment by vixalien** [comment]
I hope this also allows _enforcing_ the use of extensions as an option so that imports without extensions aren't allowed

**Comment by andrewbranch** [comment]
It does not. You might want to track #50153, and if you’re concerned with Deno, see my advice at https://github.com/microsoft/TypeScript/pull/51669#issuecomment-1349917006.

**Comment by karlhorky** [comment]
> _enforcing_ the use of extensions as an option so that imports without extensions aren't allowed

To approach this problem from another angle, you may want to consider `eslint-plugin-node` with the `node/file-extension-in-import` option:

https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/file-extension-in-import.md

This is what we've been using to enforce `.js` extensions in our TypeScript code (because we're using ESM with Node.js, which requires the extensions). Haven't tried it yet with `.ts` extensions.

**Comment by andrewbranch** [comment]
If you are compiling TS code to JS code that will run in Node, you should be using `--module nodenext` (which implies `--moduleResolution nodenext`), which requires extensions in precisely the places that Node does. No lint rule should be necessary if your reason for enforcing extensions is that you want your code to run in Node. That is built into TypeScript, and has been for about a year. #51669 is made for bundlers, and there is not a single bundler I know of (and I tested a bunch) that ever requires extensions on imports, so likewise it is not a requirement of the module resolution mode.

**Comment by ctjlewis** [comment]
> I find it funny... TS will read `import foo from "myFile"` go through the trouble of checking for (.tsx, .ts, .jsx, .js) then actually load that first match, and proceed to potentially transform that second file, but actually supporting an explicit file extension, or changing the import statement to match the about to be transformed explicit file output (that is different from input), is too far.

Yeah, it’s a lesson in design by committee for sure. I’ve argued with this team about this in these terms for long enough to know it’s like, one decision-maker who has said this is how it has to be and we are all forced to deal with it.

I wrote an entire [specification and bundler](https://github.com/tsmodule/tsmodule)  for getting “regular TS” into ESM, which this was a huge obstacle to, and I just use that.

**Comment by ctjlewis** [comment]
Just saw #51669–is this really finally fixed?!

**Comment by karlhorky** [comment]
> No lint rule should be necessary if your reason for enforcing extensions is that you want your code to run in Node. That is built into TypeScript, and has been for about a year.

We added this lint rule because we had breakages that we would only see in the production builds (because use `tsm` and `esbuild` for dev mode, which does not have a problem with lacking extensions).

The inconsistencies between bundlers allowing no extensions and `.ts` extensions and TypeScript allowing for module resolution to other file extensions like `.tsx` have been a bit of a headache to say the least:

- webpack https://github.com/webpack/webpack/issues/13252
- tsm https://github.com/lukeed/tsm/pull/33
- tsx https://github.com/esbuild-kit/tsx/issues/112
- Next.js https://github.com/vercel/next.js/issues/41961

Ideal feels like TS should just allow for `.ts` / `.tsx` / etc extensions as well and just transpile those, as an exception of the rule ["TypeScript doesn't modify JavaScript code you write"](https://github.com/microsoft/TypeScript/issues/41887#issuecomment-741968855)

**Comment by karlhorky** [comment]
But indeed, TypeScript does show the error about missing file extensions also in the IDE, which is great - thanks for the tip! I'll try applying it to those projects, maybe they are just not configured the same as this demo below:

<img width="1374" alt="Screenshot 2022-12-15 at 10 24 26" src="https://user-images.githubusercontent.com/1935696/207822177-0ecbb6f1-f967-4b50-b4ec-e28138d52be8.png">

I needed to add `"type": "module"` to my `package.json` as well as a few things to my `tsconfig.json`:

```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext"],
    "module": "NodeNext",
    "target": "ESNext",
    "moduleResolution": "NodeNext",
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "isolatedModules": true,
    "allowJs": true,
    "downlevelIteration": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "strict": true,
    "incremental": true,
    "noUncheckedIndexedAccess": true
  },
  "include": [
    "**/.eslintrc.cjs",
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    "**/*.cjs",
    "**/*.mjs"
  ],
  "exclude": ["node_modules", "build"]
}
```

**Comment by ljharb** [comment]
@andrewbranch it would be really great if TS made it easy to use native ESM *without* type module, since that package.json flag causes lots of issues with outdated tooling and also causes lots of confusion for new users.

**Comment by andrewbranch** [comment]
@ljharb can you expound on that? In modes made for Node, we’re just doing what Node requires (assuming it’s invoked with no special CLI flags altering the behavior) as far as I know. If we were to emit a `.js` file with ESM syntax, users would get this:

<img width="857" alt="image" src="https://user-images.githubusercontent.com/3277153/207928310-907253f5-1866-4c99-b339-3fbe9d8b565e.png">

<details><summary>image text</summary>

```
// a.js
import path from "path";

❯ node --version
v16.17.1
                                                                                                                                     
❯ node a.js
(node:8214) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
/Users/andrew/Developer/microsoft/eg/js/a.js:1
import path from "path";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1055:15)
    at Module._compile (node:internal/modules/cjs/loader:1090:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47

**Comment by ljharb** [comment]
@andrewbranch right - i'm asking for there to be a way to emit `.mjs` files instead (or `.cjs`, if it's CJS, i suppose).

**Comment by andrewbranch** [comment]
There is, just name your files `.mts` or `.cts`

**Comment by ljharb** [comment]
ah, ok great thanks :-)

**Comment by jeremyjacob** [comment]
See the following `tsconfig.json`:
```js
{
    "compilerOptions": {
         "moduleResolution": "bundler",
         "allowImportingTsExtensions": true
    }
}
```
